### PR TITLE
feat: autogenerated config

### LINKED ISSUE #1638: 🚀 Feature: Config auto-generate  `wrangler init --from-dash`
### Describe the solution

The configuration auto-generation (i.e. currently `wrangler.toml`) should be auto-generated from the bindings and settings in dashboard for that particular Service Worker script.

### PR DESCRIPTION
**Pull request by JacobMGEvans**
Aggregated endpoints with various points of data used in Wrangler.toml to provide a seamless experience for the user from dashboard by Using the initialization features currently built-in to Wrangler now generate a config file for the user. This will allow for a more improved Developer Experience.

related discussion #1623
resolves #1638

### CODE REVIEW

File: packages/wrangler/src/init.ts
Comment (threepointone) [inline packages/wrangler/src/init.ts:713]: since we're fetching the value for environment, maybe we shouldn't accidentally have a default here?
```suggestion
	environment: string
```

File: packages/wrangler/src/init.ts
Comment (threepointone) [inline packages/wrangler/src/init.ts:710]: this doesn't mutate, so maybe it should be called `getWorkerConfig`?
```suggestion
async function getWorkerConfig(
```

File: packages/wrangler/src/init.ts
Comment (threepointone) [inline packages/wrangler/src/init.ts:730]: What does it look like when any of these services fails/errors?

File: .changeset/sour-cobras-share.md
Comment (threepointone) [inline .changeset/sour-cobras-share.md:8]: Could you add a description here of what this PR does?

**Review by threepointone — COMMENTED** [review]
Dropped some comments, it's looking good!

File: packages/wrangler/src/init.ts
Comment (JacobMGEvans) [reply packages/wrangler/src/init.ts:730]: I was thinking about that 🤔 , would `.allSettled` make more sense or should maybe early exit with error messaging?

File: .changeset/sour-cobras-share.md
Comment (JacobMGEvans) [reply .changeset/sour-cobras-share.md:8]: I can go into a bit more detail for sure.

File: packages/wrangler/src/init.ts
Comment (petebacondarwin) [inline packages/wrangler/src/init.ts:59]: Which types file does this refer to?
Let's resolve this question in this PR and remove the comment.

File: packages/wrangler/src/init.ts
Comment (petebacondarwin) [inline packages/wrangler/src/init.ts:534]: If we are supporting environments, then perhaps we should (in the near future) add a way to specify the environment when setting the dash worker name?

Comment (petebacondarwin) [reply packages/wrangler/src/init.ts:534]: Or perhaps if there is more than one environment, we could offer a selection prompt with a list of available environments to choose from?

File: packages/wrangler/src/init.ts
Comment (petebacondarwin) [inline packages/wrangler/src/init.ts:552]: LOL! I see you are also concerned about this.

Comment (petebacondarwin) [reply packages/wrangler/src/init.ts:552]: Let's create a new issue to track that work and remove the comment here.

File: packages/wrangler/src/init.ts
Comment (petebacondarwin) [inline packages/wrangler/src/init.ts:736]: Codecov is looking for more unit tests here 😄 
It should be quite easy to add on that returns a failure from one of these API calls, right?

File: .changeset/sour-cobras-share.md
Comment (petebacondarwin) [reply .changeset/sour-cobras-share.md:8]: I don't see any additional description in the changeset file. Did you miss this?
Also the headline of this changeset doesn't seemt to be aligned with the changes in the PR.
AFAICT this PR is mainly switching over to use the non-legacy API endpoints and pulling the default environment.
Can we change the changeset to reflect this?

**Review by petebacondarwin — COMMENTED** [review]
Looking good thanks, I made a few notes that ought to be resolved before merging, please.

**Comment by Electroid** [comment]
Really cool! We should make sure that if wrangler encounters a binding type that it does not recognize, to add those to the `unsafe` binding list. Otherwise, developer may not notice, re-publish, and break their Worker.

File: packages/wrangler/src/__tests__/init.test.ts
Comment (JacobMGEvans) [reply packages/wrangler/src/__tests__/init.test.ts:2111]: Not that I could find.

File: packages/wrangler/src/init.ts
Comment (JacobMGEvans) [reply packages/wrangler/src/init.ts:59]: I'll just leave them in that file for now.

File: packages/wrangler/src/init.ts
Comment (JacobMGEvans) [reply packages/wrangler/src/init.ts:534]: I think that can be added in the future.

**Comment by JacobMGEvans** [comment]
> Really cool! We should make sure that if wrangler encounters a binding type that it does not recognize, to add those to the `unsafe` binding list. Otherwise, developer may not notice, re-publish, and break their Worker.

I can take a look at this do you know what they are called in the bindings response? Just "unsafe"?

**Comment by Electroid** [comment]
> > Really cool! We should make sure that if wrangler encounters a binding type that it does not recognize, to add those to the `unsafe` binding list. Otherwise, developer may not notice, re-publish, and break their Worker.
> 
> @Electroid I can take a look at this do you know what they are called in the bindings response? Just "unsafe"?

So let's say the API for a script returns:
```json
{"bindings":[{"name":"shhh","type":"secret_new_binding"}]}
```

Ideally, we'd want wrangler to include it in the `wrangler.toml`, even though wrangler has never heard of that binding:

```toml
[unsafe]
bindings = [{ name = "shhh", type = "secret_new_binding" }]
```

File: .changeset/sour-cobras-share.md
Comment (rozenmd) [inline .changeset/sour-cobras-share.md:9]:
```suggestion
Using the initialization features currently built-in to Wrangler
```

File: .changeset/sour-cobras-share.md
Comment (rozenmd) [inline .changeset/sour-cobras-share.md:7]: Can I get an ELI5 of what this does? from reading the code it sounds like this changeset could be

```
Makes `wrangler init`'s `--from-dash` option pull in data from Cloudflare's dashboard to generate a wrangler.toml
```
?

Comment (JacobMGEvans) [reply .changeset/sour-cobras-share.md:7]: I can add that, thanks for the suggestion.

File: packages/wrangler/src/__tests__/init.test.ts
Comment (JacobMGEvans) [reply packages/wrangler/src/__tests__/init.test.ts:2424]: I can use `.toThrowError()`? The error itself is unimportant just that it hits the error path. I believe the "Error Error" came from passing into the Error path then passed into the Mock error as well, I am not sure that is avoidable.

Comment (rozenmd) [reply packages/wrangler/src/__tests__/init.test.ts]: Ah if it's just in the mock that should be fine? Just don't want to face `Error: Error Occurred Error` as a user

File: packages/wrangler/src/init.ts
Comment (penalosa) [inline packages/wrangler/src/init.ts:77]: This should probably also include a 
```ts
    last_deployed_from?: "wrangler" | "dash" | "api"
```

File: packages/wrangler/src/init.ts
Comment (penalosa) [inline packages/wrangler/src/init.ts:517]:
```suggestion
					serviceMetaData.default_environment.environment;
```

File: packages/wrangler/src/init.ts
Comment (penalosa) [inline packages/wrangler/src/init.ts:62]: nit, but Metadata with a lowercase "d" is used in the backend—might be worth changing to that for consistency?

File: packages/wrangler/src/init.ts
Comment (penalosa) [inline packages/wrangler/src/init.ts:523]: Is the `{method: "GET"}` necessary? Is that not the default for `fetch`?

File: packages/wrangler/src/init.ts
Comment (penalosa) [inline packages/wrangler/src/init.ts:597]: I'm sure I'm missing some context here, but why is this section duplicated from above (starting line 513)? Could it be extracted into a function or something?

File: packages/wrangler/src/init.ts
Comment (penalosa) [inline packages/wrangler/src/init.ts:701]: This seems worth extracting out into an extra type (as well as extracting the ServiceMetaDataRes["environments"] type into a separate one)

File: packages/wrangler/src/init.ts
Comment (penalosa) [inline packages/wrangler/src/init.ts:711]: This seems worth extracting into a separate type (it's called `ScriptMetadata` in the backend)

File: packages/wrangler/src/init.ts
Comment (penalosa) [inline packages/wrangler/src/init.ts:734]: Is this a bug? Should it be `...(configObj?.vars ?? [])` (note the `vars`, not `var`), wrapped in `[]`? Right now it looks like only one `plain_text` or `json` binding will be stored, as each one will overwrite the next

File: packages/wrangler/src/init.ts
Comment (penalosa) [inline packages/wrangler/src/init.ts:750]: Are these enclosing `{}` needed?

File: packages/wrangler/src/init.ts
Comment (penalosa) [inline packages/wrangler/src/init.ts:848]: Is there any way this could be typed better? Since we know all the binding types, it seems this type could be constructed without `any` for better type safety (and avoiding the pluralisation bugs above)

File: packages/wrangler/src/init.ts
Comment (penalosa) [inline packages/wrangler/src/init.ts:855]: If the type is known, it might be better to use that than to use `Record<string, string>` here

File: packages/wrangler/src/init.ts
Comment (penalosa) [inline packages/wrangler/src/init.ts:875]: Can this be better typed?

**Review by penalosa — CHANGES_REQUESTED** [review]
Bunch of small things:
- Inconsistency between `.key` notation and `["key"]` notation for property accesses
- Some bugs in bindings related to pluralisation (`vars` vs `var` and `kv_namespace` vs kv_namespaces`) (at least I think so from reading the code, I could be mistaken). It looks like `plain_text`, `json`, and `kv_namespaces` are affected.
- Lots of loose types (a few instances of `any` where it's not needed), and a couple types that should probably be extracted out into their own separate type

File: packages/wrangler/src/init.ts
Comment (JacobMGEvans) [reply packages/wrangler/src/init.ts:77]: Was that merged in?

File: packages/wrangler/src/init.ts
Comment (JacobMGEvans) [reply packages/wrangler/src/init.ts:62]: Yeah that is actually one word and shouldn't be camelCase, I will change it.

File: packages/wrangler/src/init.ts
Comment (JacobMGEvans) [reply packages/wrangler/src/init.ts:523]: Not particularly.

File: packages/wrangler/src/init.ts
Comment (JacobMGEvans) [reply packages/wrangler/src/init.ts:701]: I prefer not to. It is not going to be reused and its functionally the same either way.

File: packages/wrangler/src/init.ts
Comment (JacobMGEvans) [reply packages/wrangler/src/init.ts:848]: Not really, the structure of the config is by definition unknown/any due to it potentially being any permutation of the config in Dashboard. I am open to ideas of how to get TypeScript to do it for us though, I just didn't want to end up with complex generics or manual typing that was incomplete.

File: packages/wrangler/src/init.ts
Comment (JacobMGEvans) [reply packages/wrangler/src/init.ts:855]: Fair.

File: packages/wrangler/src/init.ts
Comment (JacobMGEvans) [reply packages/wrangler/src/init.ts:875]: Environment can be any permutation of the non-inheritable config

File: packages/wrangler/src/init.ts
Comment (JacobMGEvans) [reply packages/wrangler/src/init.ts:597]: Potentially a good future refactor.

**Review by petebacondarwin — APPROVED** [review]
Approving once the tests are green.
@penalosa - good call on suggesting we tighten up the typings. It nerd sniped me into changing it and it uncovered a few fixes that were needed.

File: packages/wrangler/src/__tests__/init.test.ts
Comment (penalosa) [inline packages/wrangler/src/__tests__/init.test.ts:2088]: Is this the up to date syntax? The docs at https://developers.cloudflare.com/workers/wrangler/configuration/ suggest it's:
```ts
route = "routes" | { pattern = "*", zone_id = "ZONE_ID" } | { pattern = "*", zone_name = "ZONE_NAME" }
```

File: .changeset/sour-cobras-share.md
Comment (penalosa) [inline .changeset/sour-cobras-share.md:7]:
```suggestion
Makes `wrangler init`'s `--from-dash` option pull in data from Cloudflare's dashboard to generate a wrangler.toml file populated with configuration from an existing Worker, as well as downloading the code of the existing Worker.
```

File: packages/wrangler/src/init.ts
Comment (penalosa) [inline packages/wrangler/src/init.ts:732]: Would definitely prefer this for consistency (across the other properties too), but definitely a stylistic thing!
```suggestion
						configObj.vars = {
							...(configObj.vars ?? {}),
```

**Review by penalosa — APPROVED** [review]
Tiny things, but LGTM!

File: .changeset/sour-cobras-share.md
Comment (JacobMGEvans) [reply .changeset/sour-cobras-share.md:7]: > populated with configuration from an existing Worker, as well as downloading the code of the existing Worker.

This PR only generates the config, a previous PR generated the code

File: packages/wrangler/src/init.ts
Comment (JacobMGEvans) [reply packages/wrangler/src/init.ts:732]: For consistency I will update 😄

File: packages/wrangler/src/__tests__/init.test.ts
Comment (JacobMGEvans) [reply packages/wrangler/src/__tests__/init.test.ts:2088]: Nice catch! Thanks for the callout.

**Pull request merged**
